### PR TITLE
Add `ValueOf` type

### LINKED ISSUE #58: Proposal: ValueOf
It's a common question and not many know how to do it: https://stackoverflow.com/a/49286056/64949

```ts
type ValueOf<T> = T[keyof T];

type Foo = {a: string, b: number};

type ValueOfFoo = ValueOf<Foo>;
// string | number
```

The code is simple, but remembering the syntax is not. It would also be nice to have it as few know it's possible at all.

Real-world use-case: https://github.com/goodmind/FlowDefinitelyTyped/blob/333ce3a4273848326bfc43827eb4e7aa9a9d7441/types/react-native-vector-icons/FontAwesome5.d.ts

---

We could also consider instead adding a more powerful version that optionally supports picking the keys to get value from:

```ts
type ValueOf<T, K extends keyof T = keyof T> = T[K];

type Foo = {a: string, b: number: c: boolean};

type ValueOfFoo = ValueOf<Foo, 'a' | 'c'>;
// string | boolean
```

### PR DESCRIPTION
**Pull request by Richienb**
Fixes #58

### CODE REVIEW

**Comment by sindresorhus** [comment]
Can you read https://github.com/sindresorhus/type-fest/blob/master/.github/contributing.md and follow the steps?

File: source/basic.d.ts [hunk 1]
```diff
@@ -65,3 +65,8 @@ export interface ObservableLike {
 	subscribe(observer: (value: unknown) => void): void;
 	[Symbol.observable](): ObservableLike;
 }
+
+/**
+Matches a value in an array.
+*/
+export type ValueOf<T, ValueType extends keyof T = keyof T> = T[ValueType];
```
Comment (sindresorhus) [inline source/basic.d.ts]: This should be in its own file.

File: source/basic.d.ts [hunk 2]
```diff
@@ -65,3 +65,8 @@ export interface ObservableLike {
 	subscribe(observer: (value: unknown) => void): void;
 	[Symbol.observable](): ObservableLike;
 }
+
+/**
+Matches a value in an array.
```
[the remark this answers is not in the document]
Comment (Richienb) [reply source/basic.d.ts]: Done!

File: source/basic.d.ts (hunk 1 above)
Comment (Richienb) [reply source/basic.d.ts]: Done!

File: test-d/value-of.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import {expectType} from 'tsd';
+import {ValueOf} from '..';
+
+const value: ValueOf<[1, 2, 3]> = 1;
+
+expectType<1 | 2 | 3>(value);
```
Comment (resynth1943) [inline test-d/value-of.ts]: Mind adding a negative test here?

Comment (Richienb) [reply test-d/value-of.ts]: Done!

File: source/value-of.d.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/**
+Creates a union of all the values of the object's keys. Optionally, a key can be supplied, which limits the keys to get the values for.
+
+@example
+```
+import {ValueOf} from 'type-fest';
+
+const value: ValueOf<[1, 2, 3]> = 1
```
Comment (resynth1943) [inline source/value-of.d.ts]: It's not a blocker, but the one gripe I have with this is that this array example has the potential to cause confusion for some developers, especially with the numbers.
```suggestion
const value: ValueOf<{ key: 3 }> = 3;
```

File: test-d/value-of.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import {expectType, expectError} from 'tsd';
+import {ValueOf} from '..';
+
+const value: ValueOf<[1, 2, 3]> = 1;
```
Comment (resynth1943) [inline test-d/value-of.ts]:
```suggestion
const value: ValueOf<{ key: 3 }> = 3;
```

Why not do the same here for consistency?

File: source/value-of.d.ts (hunk 4 above)
Comment (Richienb) [reply source/value-of.d.ts]: Done!

File: test-d/value-of.ts (hunk 5 above)
Comment (Richienb) [reply test-d/value-of.ts]: Done!

**Comment by resynth1943** [comment]
@Richienb By the way, you do realise you can press the button to apply the suggested change as a commit, right?

**Comment by Richienb** [comment]
@resynth1943 That's what a did. Observe:
epico proof

**Comment by resynth1943** [comment]
> @resynth1943 That's what I did.

Yeah, this looks like a thing that GitHub doesn't do well. Might throw them an email. :heart:

File: test-d/value-of.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import {expectType, expectError} from 'tsd';
+import {ValueOf} from '..';
+
+const value: ValueOf<{a: 1; b: 2; c: 3}> = 3;
+
+expectType<1 | 2 | 3>(value);
+expectError<4>(value);
```
Comment (sindresorhus) [inline test-d/value-of.ts]: There's no test for this case:

> Optionally, a key can be supplied, which limits the keys to get the values for.

File: source/value-of.d.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+Creates a union of all the values of the object's keys. Optionally, a key can be supplied, which limits the keys to get the values for.
+
+@example
+```
+import {ValueOf} from 'type-fest';
+
+const value: ValueOf<{ a: 1; b: 2; c: 3 }> = 3;
+
+expectType<1 | 2 | 3>(value);
+```
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/value-of.d.ts]: There should also be an example for this case:

> Optionally, a key can be supplied, which limits the keys to get the values for.

File: readme.md [hunk 8]
```diff
@@ -74,6 +74,7 @@ Click the type names for complete docs.
 - [`Opaque`](source/opaque.d.ts) - Create an [opaque type](https://codemix.com/opaque-types-in-javascript/).
 - [`SetOptional`](source/set-optional.d.ts) - Create a type that makes the given keys optional.
 - [`SetRequired`](source/set-required.d.ts) - Create a type that makes the given keys required.
+- [`ValueOf`](source/value-of.d.ts) - Creates a union of all the values of the object's keys. Optionally, a key can be supplied, which limits the keys to get the values for.
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply readme.md]: It's unclear from the original description whether you can specify only one or more keys?

File: source/value-of.d.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+/**
```
Comment (sindresorhus) [inline source/value-of.d.ts]: If there has been any discussion somewhere about this type, include a link to it. For example, a discussion on the TypeScript issue tracker. - https://github.com/sindresorhus/type-fest/blob/master/.github/contributing.md

File: readme.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (Richienb) [reply readme.md]: Fixed!

File: source/value-of.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Richienb) [reply source/value-of.d.ts]: Fixed!

File: source/value-of.d.ts (hunk 9 above)
Comment (Richienb) [reply source/value-of.d.ts]: Fixed!

File: source/value-of.d.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+Creates a union of all the values of the object's keys. Optionally, a key can be supplied, which limits the keys to get the values for.
+
```
[the remark this answers is not in the document]
Comment (Richienb) [reply source/value-of.d.ts]: Fixed!

File: test-d/value-of.ts (hunk 6 above)
Comment (Richienb) [reply test-d/value-of.ts]: Fixed!

**Pull request merged**